pkg/controller/infrastructure/actuator_migrate.go

```diff
@@ -2,11 +2,52 @@ package infrastructure
 
 import (
 	"context"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
 	extensionscontroller "github.com/gardener/gardener/extensions/pkg/controller"
 	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
 )
 
-func (a *actuator) Migrate(ctx context.Context, infrastructure *extensionsv1alpha1.Infrastructure, _ *extensionscontroller.Cluster) error {
+func (a *actuator) Migrate(ctx context.Context, infrastructure *extensionsv1alpha1.Infrastructure, cluster *extensionscontroller.Cluster) error {
+	// during shoot migration, the cloud provider secret for the MCM is not being deleted preventing the migration to succeed
+	// because the secret is blocking the shoot namespace deletion.
+	//
+	// this is probably due to the reason that our MCM is provided OOT.
+
+	var (
+		deleteFinalizerName = "machine.sapcloud.io/machine-controller"
+	)
+
+	var secrets corev1.SecretList
+	err := a.client.List(ctx, &secrets, client.MatchingLabels{
+		"garden.sapcloud.io/purpose": "machineclass",
+	}, client.InNamespace(cluster.ObjectMeta.Name))
+	if err != nil {
+		return err
+	}
+
+	for _, secret := range secrets.Items {
+		secret := secret.DeepCopy()
+
+		if !strings.HasPrefix(secret.Name, "shoot--") {
+			continue
+		}
+
+		if controllerutil.ContainsFinalizer(secret, deleteFinalizerName) {
```

Comment:
Why was the check for deletionTimestamp exists removed ?

Reply:
The deletion timestamp is not set at this point in time. It's probably set during namespace deletion.

This should not be any problem though as the MCM does not seem to manage the secret anyway.